### DISCUSSION TITLE
Applying an impulse when an arrow hits a kinematic target

### DISCUSSION THREAD

**Discussion by Srekel** [jrouwe/JoltPhysics#669] [Q&A] [answered]
Hi, two-part question here.

Currently I've implemented bow arrows as dynamic bodies flying through the air. This seemed to make the most *conceptual* sense, but in retrospect I have a feeling it work work better to just simulate them manually and use raycasts to figure out what they hit. One of the problems is that they need to have quite a big shape size due to the convex radius margin. 

So my first question is - what would you recommend for arrows (primarily as it relates to performance)? This is for a typical FPS/3PS game, but I'd like to potentially have lots and lots of arrows in flight. Think a castle siege where the defenders send off a barrage.

Second question: 

The way I've implemented arrow hits then is, I have a ContactListener that collects all of the collisions during the `update`. Then, after update, I check if I hit something with health, subtract the health, and if it's <=0, I change their body from being kinematic to being dynamic. I then try to immediately apply an impulse on them. However the impulse isn't applied. It seems I must wait at least one physics update to be able to apply an impulse to the now-dynamic body. 

Would you recommend storing a list of such impulses and apply them after the next physics update? Is there a way to make it happen more directly, such as by using a StepListener?

**Comment by jrouwe** [comment, marked as answer]
> One of the problems is that they need to have quite a big shape size due to the convex radius margin.

Why don't you then reduce the convex radius (it can be zero if need be)? I would probably use a capsule shape myself because it is a simpler shape than a box and convex radius = radius so you don't need to think about it. But if you have tons and tons of arrows then it may be cheaper to simulate an arrow using a ray cast (and not give it a physics body at all).

> The way I've implemented arrow hits then is, I have a ContactListener that collects all of the collisions during the update. Then, after update, I check if I hit something with health, subtract the health, and if it's <=0, I change their body from being kinematic to being dynamic. I then try to immediately apply an impulse on them. However the impulse isn't applied. It seems I must wait at least one physics update to be able to apply an impulse to the now-dynamic body.

I'm not completely sure what you mean with ```update```, do you mean that you store the collisions from ```ContactListener::OnContactAdded/Persisted``` and then when ```PhysicsSystem::Update``` finishes change the affected bodies from ```EMotionType::Kinematic``` to ```EMotionType::Dynamic``` using ```BodyInterface::SetMotionType``` and then use ```BodyInterface::AddImpulse``` to give an impulse to the body that you just changed to Dynamic? I would expect that to work, the change from Kinematic to Dynamic is instant and the AddImpulse function will just change the velocity of the body. If you do these things from within the ```OnContactAdded/Persisted``` callbacks then I don't know what will happen as that is not supported.

**Reply by Srekel** [reply]
Ah I thought the convex radius was more or less obligatory, will look into that.

Yes I'm pretty sure that's exactly what I'm doing (not by the computer at the moment). I am using Zig bindings so I suppose it could be something mismatching there. But now that I know that that *should* work I can debug it a bit closer to see if I can figure out why it's not working.

**Reply by Srekel** [reply]
Found it - it did indeed work but the enemy AI movement code still ran for one frame after death, overwriting the impulse. It now works! 

Thanks, and I'll take your collider/raycast comments into consideration, I take it raycasts would be cheaper (which of course makes sense).

**Reply by jrouwe** [reply]
Yes, if you model your arrow as a particle (every step: v = v + g * dt, x = x + v * dt), align it to the velocity v and cast a ray between the previous position and the new position it will be a lot cheaper than doing a full blown simulation (unless you need to be able to shoot the arrow out of the sky, then the arrow is going to need some sort of collision).
